src/garage/tf/algos/rl2.py

```diff
@@ -1,15 +1,340 @@
+                * average_return: (numpy.float64)
+
+        Raises:
+            ValueError: If 'batch_idx' is not found.
+
+        """
+        concatenated_path_in_meta_batch = []
+        lengths = []
+
+        paths_by_task = collections.defaultdict(list)
+        for path in paths:
+            path['returns'] = np_tensor_utils.discount_cumsum(
+                path['rewards'], self._discount)
+            path['lengths'] = [len(path['rewards'])]
+            if 'batch_idx' in path:
+                paths_by_task[path['batch_idx']].append(path)
+            elif 'batch_idx' in path['agent_infos']:
+                paths_by_task[path['agent_infos']['batch_idx'][0]].append(path)
+            else:
+                raise ValueError('Batch idx is required for RL2 but not found')
+
+        # all path in paths_by_task[i] are sampled from task[i]
+        #
+        for path in paths_by_task.values():
+            concatenated_path = self._concatenate_paths(path)
+            concatenated_path_in_meta_batch.append(concatenated_path)
+
+        # prepare paths for inner algorithm
+        # pad the concatenated paths
+        (observations, actions, rewards, _, _, valids, lengths, env_infos,
+            agent_infos) = \
+            stack_paths(max_len=self._inner_algo.max_path_length,
+                        paths=concatenated_path_in_meta_batch)
+
+        # prepare paths for performance evaluation
+        # performance is evaluated across all paths, so each path
+        # is padded with self._max_path_length
+        (_observations, _actions, _rewards, _terminals, _, _valids, _lengths,
+            _env_infos, _agent_infos) = \
+            stack_paths(max_len=self._max_path_length,
+                        paths=paths)
+
+        ent = np.sum(self._policy.distribution.entropy(agent_infos) *
+                     valids) / np.sum(valids)
+
+        undiscounted_returns = log_performance(
+            itr, TrajectoryBatch.from_trajectory_list(self._env_spec, paths),
+            self._inner_algo.discount)
+
+        tabular.record('Entropy', ent)
+        tabular.record('Perplexity', np.exp(ent))
+
+        # all paths in each meta batch is stacked together
+        # shape: [meta_batch, max_path_length * episoder_per_task, *dims]
+        # per RL^2
+        concatenated_path = dict(observations=observations,
+                                 actions=actions,
+                                 rewards=rewards,
+                                 valids=valids,
+                                 lengths=lengths,
+                                 baselines=np.zeros_like(rewards),
+                                 agent_infos=agent_infos,
+                                 env_infos=env_infos,
+                                 paths=concatenated_path_in_meta_batch,
+                                 average_return=np.mean(undiscounted_returns))
+
+        return concatenated_path
+
+    def _concatenate_paths(self, paths):
```

Comment:
@krzentner can this not be achieved by your recent TrajectoryBatch helpers?

Reply:
I'm pretty sure this function can either use that, or `tensor_utils.concat_tensor_dict_list` to be 80% shorter.

Reply:
fixed